File: rest-service/src/main/java/com/example/restservice/service/SubscriptionService.java
package com.example.restservice.service;

import com.example.restservice.model.Subscription;

import java.util.List;

/**
 * Interfața care definește operațiile disponibile pentru gestionarea abonamentelor în sistem.
 * Aceste operații includ obținerea tuturor abonamentelor, obținerea unui abonament după ID,
 * crearea, actualizarea și ștergerea abonamentelor, precum și obținerea abonamentelor pentru un anumit utilizator.
 */
public interface SubscriptionService {
    /**
     * Obține o listă cu toate abonamentele din sistem.
     *
     * @return Lista cu abonamente
     */
    List<Subscription> getAllSubscriptions();

    /**
     * Obține un abonament după ID-ul specificat.
     *
     * @param id ID-ul abonamentului căutat
     * @return Abonamentul cu ID-ul specificat sau null dacă nu există
     */
    Subscription getSubscriptionById(Long id);

    /**
     * Creează un abonament nou în sistem.
     *
     * @param subscription Abonamentul care urmează să fie creat
     * @return Abonamentul creat
     */
    Subscription createSubscription(Subscription subscription);

    /**
     * Actualizează un abonament existent în sistem.
     *
     * @param id                 ID-ul abonamentului care urmează să fie actualizat
     * @param updateSubscription Abonamentul actualizat
     * @return Abonamentul actualizat sau null dacă nu există un abonament cu ID-ul specificat
     */
    Subscription updateSubscription(Long id, Subscription updateSubscription);

    /**
     * Șterge un abonament existent din sistem.
     *
     * @param id ID-ul abonamentului care urmează să fie șters
     */
    void deleteSubscription(Long id);

    /**
     * Obține o listă cu abonamentele asociate unui anumit utilizator.
     *
     * @param userId ID-ul utilizatorului pentru care se caută abonamentele
     * @return Lista cu abonamentele asociate utilizatorului specificat
     */
    List<Subscription> getSubscriptionsByUserId(Long userId);
}

File: front-site/src/app/service/subscription.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { User } from './user.service';
import { Category } from './category.service';

// Definirea interfeței Subscription
export interface Subscription {
  subscriptionId?: number;
  user: User;
  category: Category;
  startDate: string; // LocalDate este reprezentat ca string în JSON
  amountPaid: number;
}



@Injectable({
  providedIn: 'root'
})
export class SubscriptionService {
  private baseUrl = 'http://localhost:8080/subscriptions';

  constructor(private http: HttpClient) { }

  // Metodă pentru obținerea tuturor abonamentelor
  getAllSubscriptions(): Observable<Subscription[]> {
    return this.http.get<Subscription[]>(this.baseUrl);
  }

  // Metodă pentru obținerea unui abonament după ID
  getSubscriptionById(id: number): Observable<Subscription> {
    return this.http.get<Subscription>(`${this.baseUrl}/${id}`);
  }

  // Metodă pentru crearea unui abonament nou
  createSubscription(subscription: Subscription): Observable<Subscription> {
    return this.http.post<Subscription>(this.baseUrl, subscription);
  }

  // Metodă pentru actualizarea unui abonament existent
  updateSubscription(id: number, subscription: Subscription): Observable<Subscription> {
    return this.http.put<Subscription>(`${this.baseUrl}/${id}`, subscription);
  }

  // Metodă pentru ștergerea unui abonament după ID
  deleteSubscription(id: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/${id}`);
  }

  // Metodă pentru obținerea abonamentelor unui utilizator după ID-ul utilizatorului
  getSubscriptionsByUserId(userId: number): Observable<Subscription[]> {
    return this.http.get<Subscription[]>(`${this.baseUrl}/getSubscriptionsByUserId/${userId}`);
  }
}
